File: number_theory/04_binary_exponentiation.py


def binary_expn(a : int, n : int) -> int:
	if( n == 0 ): return 1
	if( n == 1 ): return a
	res = 1
	while(n):
		if( n % 2 ):
			res *= a
			n -= 1
		a *= a
		n //= 2
	return res

base, power = map(int, input("Enter base and power : ").split(" "))
print(binary_expn(base,power))

File: number_theory/04_binary_exponentiation.cpp
#include<iostream>
using namespace std;

long long binary_expn(int a, int n){
	if( n == 0 ) return 1;
	if( n == 1 ) return a;
	long long res = 1;
	while(n){
		if( n % 2 ){
			res *= a;
			n--;
		}
		a *= a;
		n /= 2;
	}
	return res;
}

int main(){
	int base, power;
	cout << "Enter base and power : ";
	cin >> base >> power;
	cout << binary_expn(base,power) << endl;
	return 0;
}